### DISCUSSION TITLE
Compute and save loss on each training+val sample individually?

### DISCUSSION THREAD

**Discussion by nick-lavigne-ursaspace** [ultralytics/ultralytics#6946] [Q&A] [answered]
Does anyone know if it is currently possible or easily do-able to compute the loss values on each individual sample point in the training and val set and then save them off for review? I want to review my training data based on which samples are most difficult for the model at the end of training.

**Comment by pderrenger** [comment]
@nick-lavigne-ursaspace currently, YOLOv8 does not provide a built-in feature to save losses for each individual sample directly during training or validation. However, you can modify the training loop to log the loss for each sample.

To achieve this, you would need to edit the code where the batch loss is computed and log the individual losses. This would involve iterating over each sample in the batch, computing the loss, and then saving it to a file or a data structure for later analysis.

Keep in mind that this could slow down your training process significantly, especially for large datasets, as it increases the computational overhead. Also, you would need to ensure that the data structure or file handling is efficient enough to handle the potentially large amount of data generated.

If you're comfortable with modifying the code, you can start by looking into the training loop within the `train.py` file (or its equivalent in YOLOv8) and make the necessary adjustments to capture and record the loss per sample. Remember to handle both the training and validation phases if you need this data for both.

For a more detailed guide on how to implement this, you might want to check out the documentation or seek further assistance from the community. 📘🛠️

**Reply by nick-lavigne-ursaspace** [reply]
Hi @pderrenger thanks for getting back to me about this! 

To clarify, I'm not looking to compute the loss on each sample during training, just one time on every sample in the training and validation set. Like you said I wouldn't want to incur the computational costs of doing this on every training iteration (although I could see the benefit of doing it at a regular interval for big training jobs). 

There is already the option to save model predictions as a txt file similar to how the labels are saved. I think the simplest way for me to achieve my goal would be to predict on each sample in the training and validation sets and then compute the loss on these outputs. 

Could you point me to where in the code base the loss function is that I could import to do this calculation after the predictions are already made?

**Reply by pderrenger** [reply]
Hi @nick-lavigne-ursaspace, I understand your requirement better now. You're looking to evaluate the loss on each sample post-training to analyze the difficulty level of the samples for the model.

To compute the loss for each sample after predictions have been made, you would typically use the loss functions that were employed during training. In YOLOv8, the loss computation is integrated into the training loop, and the specific loss functions depend on the tasks (object detection, segmentation, etc.).

The loss functions are usually defined in the `models` directory within the codebase. For example, in YOLOv5, the loss for object detection is computed in `models/yolo.py` within the `forward` method of the `Detect` class. YOLOv8 will have a similar setup, but the exact file and class names may differ.

Here's a general approach to what you're looking to do:

1. Load the trained model using the YOLO class.
2. Iterate over your dataset, passing each image through the model to get predictions.
3. Use the corresponding loss function to compute the loss between predictions and ground truth labels.
4. Save the computed loss values for further analysis.

Here's a simplified example in Python:

```python
from ultralytics import YOLO, load_dataset
from pathlib import Path

# Load the trained model
model = YOLO('path/to/your/trained/model.pt')

# Load your dataset
dataset = load_dataset('path/to/your/dataset.yaml')
dataloader = dataset.dataloader(batch_size=1, imgsz=model.imgsz, shuffle=False)

# Placeholder for losses
losses = []

# Iterate over the dataset
for batch_i, (imgs, targets, paths, _) in enumerate(dataloader):
    imgs = imgs.to(model.device)
    targets = targets.to(model.device)
    
    # Get predictions
    with torch.no_grad():
        pred = model(imgs)
    
    # Compute loss
    loss = model.compute_loss(pred, targets)  # This method will vary based on your model's implementation
    losses.append((Path(paths[0]).name, loss.item()))

# Save losses for analysis
with open('losses.txt', 'w') as f:
    for path, loss in losses:
        f.write(f'{path}: {loss}\n')
```

Please note that you'll need to adapt the code to fit the exact implementation of your YOLOv8 model, especially the `compute_loss` method, which may not be directly accessible or may require certain inputs that are only available during training. You might need to refactor the loss computation code to make it usable outside the training loop.

If you're not sure where to find the loss functions or how to use them, I recommend checking the YOLOv8 documentation or reaching out to the community for more specific guidance. 📚👨‍💻

**Reply by nick-lavigne-ursaspace** [reply]
Thank you, I'm getting close to a solution here. I think I need to use v8DetectionLoss but it's a little unclear to me how to format the inputs. Here is what I have:
```python
import os
import yaml

from ultralytics import YOLO
from ultralytics.utils.loss import v8DetectionLoss
from ultralytics.data.dataset import YOLODataset
from torch.utils.data import DataLoader


model_filename = "..."
data_yaml_filename = "..."

with open(data_yaml_filename, 'r') as f:
    data = yaml.safe_load(f)

train_data_path = os.path.join(data['path'], 'train')
val_data_path = os.path.join(data['path'], 'val')

model = YOLO(model_filename)
loss = v8DetectionLoss(model.model)

train_dataset = YOLODataset(img_path=train_data_path,
                            data=data)
train_dataloader = DataLoader(train_dataset,
                              batch_size=1,
                              shuffle=False)

for batch in train_dataloader:
    pred = model(batch['im_file'], imgsz=1280)
    loss(pred, batch) # currently errors here
``` 

The `v8DetectionLoss` seems to require the raw tensor output of the model, not the formatted results object. Do you know how I can obtain that?

**Reply by pderrenger** [reply]
@nick-lavigne-ursaspace you're correct that the `v8DetectionLoss` function requires the raw predictions from the model, not the post-processed results. To obtain the raw predictions, you need to pass your images directly through the model without the post-processing that converts them into a `Results` object.

Here's how you can modify your code to get the raw predictions:

```python
import torch
from ultralytics import YOLO
from ultralytics.data.dataset import LoadImagesAndLabels
from torch.utils.data import DataLoader

# Load the model and loss function
model_filename = "path/to/your/model.pt"
data_yaml_filename = "path/to/your/data.yaml"
model = YOLO(model_filename).to('cuda')
loss_fn = model.model.loss_fn

# Load the dataset
with open(data_yaml_filename) as f:
    data = yaml.safe_load(f)
train_dataset = LoadImagesAndLabels(data['train'], img_size=model.imgsz, batch_size=1, augment=False)
train_dataloader = DataLoader(train_dataset, batch_size=1, shuffle=False)

# Compute loss for each sample
for batch_i, (imgs, targets, paths, _) in enumerate(train_dataloader):
    imgs = imgs.to('cuda')
    targets = targets.to('cuda')
    
    # Forward pass
    pred = model.model(imgs)  # Raw predictions
    
    # Compute loss
    loss, loss_items = loss_fn(pred, targets)
    print(f"Image: {paths[0]}, Loss: {loss.item()}")
```

In this code snippet, `model.model` is the underlying PyTorch model that outputs the raw predictions, and `loss_fn` is the loss function associated with the model. The `LoadImagesAndLabels` class is used to load the dataset, and the `DataLoader` is used to iterate over the dataset.

Please note that the `LoadImagesAndLabels` class and the loss function may have different names or require different parameters in YOLOv8. You'll need to adjust the code accordingly based on the actual implementation in YOLOv8.

Also, ensure that the `targets` tensor is correctly formatted to match the expected input of the loss function. The `targets` tensor typically contains information about the ground truth bounding boxes, including the class labels and box coordinates.

By iterating over the dataset and computing the loss for each sample, you can analyze which samples are most difficult for the model. Remember to handle the data and model device placement (`'cuda'` or `'cpu'`) appropriately based on your setup.

**Reply by nick-lavigne-ursaspace** [reply, marked as answer]
I think I've got it! At least, this seems to loop through the training and validation sets and successfully computes the loss function on each of them. 

```python
import os
import yaml

from ultralytics import YOLO
from ultralytics.data.dataset import YOLODataset
from ultralytics.data import build_dataloader


model_filename = "..."
data_yaml_filename = "...l"

model = YOLO(model_filename)
loss = model.model.criterion

with open(data_yaml_filename, 'r') as f:
    data = yaml.safe_load(f)

train_data_path = os.path.join(data['path'], 'train')
val_data_path = os.path.join(data['path'], 'val')


train_dataset = YOLODataset(
    img_path=train_data_path,
    data=data,
    imgsz=1280,
    augment='val'
)
train_dataloader = build_dataloader(train_dataset, workers=8, batch=1, shuffle=False)

val_dataset = YOLODataset(
    img_path=val_data_path,
    data=data,
    imgsz=1280,
    augment='val'
)
val_dataloader = build_dataloader(val_dataset, workers=8, batch=1, shuffle=False)


for batch in train_dataloader:
    pred = model.model(batch['img']/255)
    print(loss(pred, batch))

for batch in val_dataloader:
    pred = model.model(batch['img'] / 255)
    print(loss(pred, batch))
``` 

The only part I wasn't sure about was where I had to divide the `batch['img']` by 255 to convert from uint8 0-255 to float 0-1. I would have assumed the data loader would do that for me any idea why this is the case? Is there something else I'm missing?

**Reply by pderrenger** [reply]
@nick-lavigne-ursaspace great to hear you've made progress! Your approach to loop through the datasets and compute the loss is correct.

Regarding the image normalization (dividing by 255), this is typically handled during preprocessing. If the `YOLODataset` does not perform this normalization, you would indeed need to manually adjust the pixel values to the 0-1 range expected by the model.

It's possible that the `augment` parameter in `YOLODataset` is set to `'val'`, which might not include the normalization step. You can check the dataset and dataloader implementation to confirm if normalization is included, or you can continue to manually normalize as you've done.

Keep up the good work, and if you have further questions, don't hesitate to ask! 👍
